compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVerifyVectorSizeLegality.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <climits>
+#include <numeric>
+
+#include "iree/compiler/Codegen/LLVMCPU/PassDetail.h"
+#include "iree/compiler/Codegen/LLVMCPU/Passes.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "llvm/Support/Alignment.h"
+#include "llvm/Support/MathExtras.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir::iree_compiler {
+
+namespace {
+struct LLVMCPUVerifyVectorSizeLegalityPass
+    : LLVMCPUVerifyVectorSizeLegalityBase<LLVMCPUVerifyVectorSizeLegalityPass> {
+  LLVMCPUVerifyVectorSizeLegalityPass(int64_t ratio) { this->ratio = ratio; }
+
+  void runOnOperation() override;
+};
+} // namespace
+
+static int64_t getTotalSizeInBytes(Type elemType, ArrayRef<int64_t> shape) {
+  // We can't query bitwidth for some types (e.g., index type). For those
+  // cases, we assume that they are 64 bits because most of modern systems are
+  // 64-bit.
+  int64_t elemBitWidth = 64;
+  if (elemType.isIntOrFloat()) {
+    elemBitWidth = elemType.getIntOrFloatBitWidth();
+  }
+  int64_t size = std::accumulate(shape.begin(), shape.end(), 1LL,
+                                 std::multiplies<int64_t>{});
+  size *= elemBitWidth;
+  size = llvm::divideCeil(size, 8);
+  return size;
+}
+
+void LLVMCPUVerifyVectorSizeLegalityPass::runOnOperation() {
+  FunctionOpInterface funcOp = getOperation();
+  // Use 64 bits as target hardware vector size if the native_vector_size is not
+  // present.
+  int64_t maxVectorSizeInBytes = 8;
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(funcOp);
+  if (targetAttr) {
+    auto nativeVectorSizeAttr =
+        getConfigIntegerAttr(targetAttr, "native_vector_size");
+    if (nativeVectorSizeAttr) {
+      maxVectorSizeInBytes = nativeVectorSizeAttr->getInt();
+    }
+  }
+  constexpr int64_t kInt64Max = std::numeric_limits<int64_t>::max();
+  if (maxVectorSizeInBytes <= kInt64Max / this->ratio) {
+    maxVectorSizeInBytes *= this->ratio;
+  } else {
+    maxVectorSizeInBytes = kInt64Max;
+  }
```

Comment:
I feel that perhaps this case should be an error - that would be quite a large vector 😅 And, IIUC, the goal is to prevent large vectors ever appearing.